core/src/main/java/org/elasticsearch/search/suggest/completion/context/ContextMappings.java

```diff
@@ -0,0 +1,322 @@
+        }
+    }
+
+    /**
+     * @return number of context mappings
+     * held by this instance
+     */
+    public int size() {
+        return contextMappings.size();
+    }
+
+    /**
+     * Returns a context mapping by its name
+     */
+    public ContextMapping<?> get(String name) {
+        ContextMapping contextMapping = contextNameMap.get(name);
+        if (contextMapping == null) {
+            throw new IllegalArgumentException("Unknown context name[" + name + "], must be one of " + contextNameMap.size());
+        }
+        return contextMapping;
+    }
+
+    /**
+     * Adds a context-enabled field for all the defined mappings to <code>document</code>
+     * see {@link org.elasticsearch.search.suggest.completion.context.ContextMappings.TypedContextField}
+     */
+    public void addFields(ParseContext.Document document, String name, String input, int weight, Map<String, Set<CharSequence>> contexts) {
+        for (int typeId = 0; typeId < contextMappings.size(); typeId++) {
+            ContextMapping<?> mapping = contextMappings.get(typeId);
+            Set<CharSequence> ctxs = contexts.get(mapping.name());
+            if (ctxs == null) {
+                ctxs = new HashSet<>();
+            }
+            document.add(new TypedContextField(name, typeId, input, weight, ctxs, document, mapping));
+        }
+    }
+
+    /**
+     * Field prepends context values with a suggestion
+     * Context values are associated with a type, denoted by
+     * a type byte, which is prepended to the context value.
+     *
+     * Every defined context mapping yields a unique type id
+     * for all its context values
+     *
+     * The type, context and suggestion values are encoded as follows:
+     * <p>
+     *     TYPE_ID | CONTEXT_VALUE | CONTEXT_SEP | SUGGESTION_VALUE
+     * </p>
+     *
+     * Field can also use values of other indexed fields as contexts
+     * at index time
+     */
+    private static class TypedContextField extends ContextSuggestField {
+        private final int typeId;
+        private final Set<CharSequence> contexts;
+        private final ParseContext.Document document;
+        private final ContextMapping<?> mapping;
+
+        public TypedContextField(String name, int typeId, String value, int weight, Set<CharSequence> contexts,
+                                 ParseContext.Document document, ContextMapping mapping) {
+            super(name, value, weight);
+            this.typeId = typeId;
+            this.contexts = contexts;
+            this.document = document;
+            this.mapping = mapping;
+        }
+
+        @Override
+        protected Iterable<CharSequence> contexts() {
+            contexts.addAll(mapping.parseContext(document));
+            final Iterator<CharSequence> contextsIterator = contexts.iterator();
+            final CharsRefBuilder scratch = new CharsRefBuilder();
+            scratch.append(((char) typeId));
```

Comment:
Hmm, how large is typeId allowed to be (how many unique context types are allowed)?  It's an int, but we cast to char here, and javadocs say it's a byte :)  Can we check this and throw a clear exc if user tries to make too many?